src/l1_fetcher.rs

```diff
@@ -131,8 +126,31 @@ impl L1Fetcher {
             }
         });
 
+        // If an `end_block` was supplied we shouldn't poll for newer blocks.
+        if end_block.is_some() {
+            disable_polling = true;
+        }
+
+        let end_block_number = end_block.unwrap_or(
+            self.provider
+                .get_block(BlockNumber::Latest)
+                .await?
+                .unwrap()
+                .number
+                .unwrap(),
+        );
+
         let mut latest_l2_block_number = U256::zero();
-        while current_l1_block_number <= end_block_number {
+        loop {
+            if disable_polling && current_l1_block_number > end_block_number {
+                tracing::info!("Successfully reached end block. Shutting down...");
+                tx_handle.abort();
+                parse_handle.abort();
```

Comment:
I think we should rather drop the `sink` and let the tasks later in the chain naturally exit. This would avoid loosing "last bits of work". Otherwise there's race between different threads.